Write Read, Patch & Replace ReplicaSet Status +3 endpoints

riaankleinhans opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- replaceAppsV1NamespacedReplicaSetStatus
- readAppsV1NamespacedReplicaSetStatus
- patchAppsV1NamespacedReplicaSetStatus

**Which issue(s) this PR fixes:**
Fixes #100863
**Testgrid Link:**
[Test](https://k8s-testgrid.appspot.com/sig-release-master-blocking#gce-cos-master-default&&width=5&include-filter-by-regex=should%20validate%20Replicaset%20Status%20endpoints&graph-metrics=test-duration-minutes)

**Special notes for your reviewer:**
Adds +3 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

Riaankl commented:
/triage accepted

Riaankl commented:
/assign @soltysh @johnbelamaric @smarterclayton

dims commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +534,138 @@ func testRSLifeCycle(f *framework.Framework) {
+		"pod":  WebserverImageName,
+	}
+	labelSelector := labels.SelectorFromSet(rsPodLabels).String()
+	framework.Logf("labels: %v", labelSelector)
```
is this needed? if so, could there be a longer description message?

heyste replied:
Removed.

dims commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +534,138 @@ func testRSLifeCycle(f *framework.Framework) {
+		return err
+	})
+	framework.ExpectNoError(err, "Failed to update status. %v", err)
+	framework.Logf("updatedStatus.Conditions: %#v", updatedStatus.Status.Conditions)
```
need this log?

heyste replied:
If the test was to flake at some point the idea was to have a reference point to help debug what status conditions we have at this point of the test and what we validate with the watch below.

dims commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +534,138 @@ func testRSLifeCycle(f *framework.Framework) {
+				}
+			}
+		}
+		framework.Logf("Observed event: %+v", event.Type)
```
need this log?

heyste replied:
The above logging was due to [feedback](https://github.com/kubernetes/kubernetes/pull/96485#issuecomment-732335344) when trying to debug a test flake that uses a similar code snippet. Extra logging was added [back into the test](https://github.com/kubernetes/kubernetes/pull/97137) to address the concerns raised by @spiffxp.

dims commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +534,138 @@ func testRSLifeCycle(f *framework.Framework) {
+
+	patchedReplicaSet, err := rsClient.Patch(context.TODO(), rsName, types.MergePatchType, payload, metav1.PatchOptions{}, "status")
+	framework.ExpectNoError(err, "Failed to patch status. %v", err)
+	framework.Logf("Patched status conditions: %#v", patchedReplicaSet.Status.Conditions)
```
one more

heyste replied:
see comment above for line 602

dims commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +534,138 @@ func testRSLifeCycle(f *framework.Framework) {
+				}
+			}
+		}
+		framework.Logf("Observed event: %+v", event.Type)
```
one more

heyste replied:
see comment above for line 629

ravisantoshgudimetla commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +534,138 @@ func testRSLifeCycle(f *framework.Framework) {
+	err = e2epod.VerifyPodsRunning(c, ns, "sample-pod", false, replicas)
+	framework.ExpectNoError(err, "Failed to create pods: %s", err)
+
+	ginkgo.By("Getting /status")
```
Is the goal here just to hit `/status` endpoint? I ask because to get status sub-resource, `.Get()` method should have been enough similar to the approach you followed below.

heyste replied:
@ravisantoshgudimetla I'm on the same team has @Riaankl and wrote this test to be part of conformance.

Yes, the goal is to hit the `/status` endpoint. The `.Get()` method can't access the status sub-resource, so it requires the dynamic client.

ravisantoshgudimetla commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +534,138 @@ func testRSLifeCycle(f *framework.Framework) {
+	framework.ExpectNoError(err, "Failed to unmarshal JSON bytes to a replicaset object type")
+	framework.Logf("Replicaset %s has Conditions: %v", rsName, rsStatus.Status.Conditions)
+
+	ginkgo.By("updating the Replicaset Status")
```
We usually expect the controller to update the status not by other actor. So, I am wondering if we are just hitting this to increase the coverage where as we in reality this cannot and shouldn't happen

ravisantoshgudimetla replied:
I am wondering if you can mimic the patch by modifying the spec and watch if the status subresource has been patched instead of manually updating the status

heyste replied:
The process of checking `/status` endpoints has gone through a number of discussions in conformance meetings for [Service Status endpoints](https://github.com/kubernetes/kubernetes/issues/94867) and [API Service Status](https://github.com/ii/kubernetes/blob/apiservice-status-endpoints/test/e2e/apimachinery/aggregator.go#L546-L670). 

I'm trying to have a clean common pattern for testing any `/status` endpoint no matter the resource.

ravisantoshgudimetla reviewed (changes requested):
Thank you for working on this @Riaankl. I have few questions around the api resources being used directly. Please look at my comments.

ravisantoshgudimetla reviewed (approved):
Thank you @Riaankl and @heyste for working on this PR.

/lgtm

Riaankl commented:
/assign @soltysh @mattfarina
This e2e test is to clean up the last few ReplicaSet endpoints.
Following the same pattern as previous Mergde Conformance tests
It has been reviewed, can we please get a `/approve` to get it on the test grid.

heyste commented:
/test pull-kubernetes-node-e2e-containerd

x2 flakes 
- `Node Tests`
- `E2e Node  Suite: [sig-node] Summary API [NodeConformance] when querying  /stats/summary should report resource usage through the stats api`

heyste commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +535,135 @@ func testRSLifeCycle(f *framework.Framework) {
+					framework.Logf("Found replicaset %v in namespace %v with labels: %v annotations: %v & Conditions: %v", rs.ObjectMeta.Name, rs.ObjectMeta.Namespace, rs.ObjectMeta.Labels, rs.Annotations, rs.Status.Conditions)
+					return found, nil
+				}
+				framework.Logf("Observed replicaset %v in namespace %v with annotations: %v & Conditions: %v", rs.ObjectMeta.Name, rs.ObjectMeta.Namespace, rs.Annotations, rs.Status.Conditions)
+			}
+		}
+		object := strings.Split(fmt.Sprintf("%v", event.Object), "{")[0]
+		framework.Logf("Observed %v event: %+v", object, event.Type)
+		return false, nil
```
@ravisantoshgudimetla Thanks for the earlier review. I've updated this section as there was a potential bug where I had an extra `return`. Also, line 627-8 includes the object for the event now.

Can I get another lgtm? TIA

heyste commented on test/e2e/apps/replica_set.go:
```diff
@@ -524,3 +535,135 @@ func testRSLifeCycle(f *framework.Framework) {
+					framework.Logf("Found replicaset %v in namespace %v with labels: %v annotations: %v & Conditions: %v", rs.ObjectMeta.Name, rs.ObjectMeta.Namespace, rs.ObjectMeta.Labels, rs.Annotations, cond)
+					return found, nil
+				}
+				framework.Logf("Observed replicaset %v in namespace %v with annotations: %v & Conditions: %v", rs.ObjectMeta.Name, rs.ObjectMeta.Namespace, rs.Annotations, cond)
+			}
+		}
+		object := strings.Split(fmt.Sprintf("%v", event.Object), "{")[0]
+		framework.Logf("Observed %v event: %+v", object, event.Type)
```
@ravisantoshgudimetla same as line 624-9

heyste committed:
Create e2e Replicaset status lifecycle test

e2e test validates the following 3 extra endpoints
-  replaceAppsV1NamespacedReplicaSetStatus
-  readAppsV1NamespacedReplicaSetStatus
-  patchAppsV1NamespacedReplicaSetStatus

ravisantoshgudimetla reviewed (approved):
/lgtm

smarterclayton commented:
/approve
